framework/test/jest/unit/specs/modules/chain/synchronizer/fast_chain_switching_mechanism/fast_chain_switching_mechanism.spec.js

```diff
@@ -14,149 +14,866 @@
+	let dposModuleMock;
+	let exceptions;
+	let loggerMock;
+	let storageMock;
 
-		const storageMock = {
+	beforeEach(() => {
+		loggerMock = {
+			info: jest.fn(),
+			debug: jest.fn(),
+			error: jest.fn(),
+			trace: jest.fn(),
+		};
+		storageMock = {
 			entities: {
+				TempBlock: {
+					get: jest.fn(),
+					truncate: jest.fn(),
+					isEmpty: jest.fn(),
+				},
 				Block: {
-					getLastBlock: jest.fn(),
+					get: jest.fn(),
+					getOne: jest.fn(),
+					begin: jest.fn(),
+				},
+				Account: {
+					get: jest.fn(),
+				},
+				ChainMeta: {
+					getKey: jest.fn(),
 				},
 			},
 		};
-		const dposMock = {
-			getForgerPublicKeysForRound: jest.fn().mockReturnValue([]),
-		};
-		const slotsMock = {
-			getSlotNumber: jest.fn(),
-			calcRound: jest.fn(),
-		};
-		const syncParams = {
+		channelMock = new ChannelMock();
+
+		slots = new Slots({
+			epochTime: constants.EPOCH_TIME,
+			interval: constants.BLOCK_TIME,
+			blocksPerRound: constants.ACTIVE_DELEGATES,
+		});
+
+		bftModule = new BFT({
+			storage: storageMock,
+			logger: loggerMock,
+			activeDelegates: constants.ACTIVE_DELEGATES,
+			startingHeight: 1,
+		});
+
+		blocksModule = new Blocks({
+			logger: loggerMock,
+			storage: storageMock,
+			slots,
+			genesisBlock: genesisBlockDevnet,
+			sequence: new Sequence(),
+			interfaceAdapters: {
+				transactions: new TransactionInterfaceAdapter(),
+			},
+			blockReceiptTimeout: constants.BLOCK_RECEIPT_TIMEOUT,
+			loadPerIteration: 1000,
+			maxPayloadLength: constants.MAX_PAYLOAD_LENGTH,
+			maxTransactionsPerBlock: constants.MAX_TRANSACTIONS_PER_BLOCK,
+			activeDelegates: constants.ACTIVE_DELEGATES,
+			rewardDistance: constants.REWARDS.DISTANCE,
+			rewardOffset: constants.REWARDS.OFFSET,
+			rewardMileStones: constants.REWARDS.MILESTONES,
+			totalAmount: constants.TOTAL_AMOUNT,
+			blockSlotWindow: constants.BLOCK_SLOT_WINDOW,
+		});
+		blocksModule.getTempBlocks = jest.fn();
+
+		blockProcessorV2 = new BlockProcessorV2({
+			blocksModule,
+			bftModule,
+			dposModule: dposModuleMock,
+			logger: loggerMock,
+			constants,
+			exceptions,
+		});
+
+		processorModule = new Processor({
 			channel: channelMock,
-			blocks: blocksMock,
 			storage: storageMock,
-			slots: slotsMock,
-			dpos: dposMock,
-			activeDelegates,
-		};
+			blocksModule,
+			logger: loggerMock,
+		});
+		processorModule.processValidated = jest.fn();
+		processorModule.validate = jest.fn();
+		processorModule.deleteLastBlock = jest.fn();
+		processorModule.register(blockProcessorV2);
+
+		fastChainSwitchingMechanism = new FastChainSwitchingMechanism({
+			storage: storageMock,
+			logger: loggerMock,
+			channel: channelMock,
+			slots,
+			blocks: blocksModule,
+			bft: bftModule,
+			processor: processorModule,
+			dpos: dposModuleMock,
+			activeDelegates: constants.ACTIVE_DELEGATES,
+		});
+	});
+
+	describe('async run()', () => {
+		const aPeer = '127.0.0.1:5000';
+		let aBlock;
+
+		function checkApplyPenaltyAndRestartIsCalled(
```

Comment:
I think in mocha test there were several function which calls expect, and it was hard to tell where it was actually throwing error. what do you think?
Also, defining in arrow function is fine IMO `const checkApplyPenaltyAndRestartIsCalled = () => {}`

Reply:
Yeah, @yatki and I are discussing how to do this in a clearer way. The thing is those expects are repeatedly called on multiple test cases and I didn't want to clone the code multiple times.